### ISSUE TITLE
in DCGAN example, Why do we need to make netD inference twice?

### ISSUE THREAD

**Issue by DreamChaserMXF** [pytorch/examples#584]
To my understanding, the two lines nearly have the same functionality (same input, same output) except in the first inference `fake` is detached.

https://github.com/pytorch/examples/blob/1de2ff9338bacaaffa123d03ce53d7522d5dcc2e/dcgan/main.py#L229

https://github.com/pytorch/examples/blob/1de2ff9338bacaaffa123d03ce53d7522d5dcc2e/dcgan/main.py#L241

Is it necessary to make netD inference twice with the same input?
What if we re-use the output of the first inference (with fake input) to calculate `errG`?

instead of the original codes:
```
        ############################
        # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
        ###########################
        # train with real
        netD.zero_grad()
        real_cpu = data[0].to(device)
        batch_size = real_cpu.size(0)
        label = torch.full((batch_size,), real_label, device=device)

        output = netD(real_cpu)
        errD_real = criterion(output, label)
        errD_real.backward()
        D_x = output.mean().item()

        # train with fake
        noise = torch.randn(batch_size, nz, 1, 1, device=device)
        fake = netG(noise)
        label.fill_(fake_label)
        output = netD(fake.detach())
        errD_fake = criterion(output, label)
        errD_fake.backward()
        D_G_z1 = output.mean().item()
        errD = errD_real + errD_fake
        optimizerD.step()

        ############################
        # (2) Update G network: maximize log(D(G(z)))
        ###########################
        netG.zero_grad()
        label.fill_(real_label)  # fake labels are real for generator cost
        output = netD(fake)
        errG = criterion(output, label)
        errG.backward()
        D_G_z2 = output.mean().item()
        optimizerG.step()
```

can we re-write them as following?
```
        ############################
        # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
        ###########################
        # train with real
        netD.zero_grad()
        real_cpu = data[0].to(device)
        batch_size = real_cpu.size(0)
        label = torch.full((batch_size,), real_label, device=device)

        real_output = netD(real_cpu)
        errD_real = criterion(real_output, label)
        errD_real.backward()
        D_x = real_output.mean().item()

        # train with fake
        noise = torch.randn(batch_size, nz, 1, 1, device=device)
        fake = netG(noise)
        label.fill_(fake_label)
        # output = netD(fake.detach())
        fake_output = netD(fake)
        errD_fake = criterion(fake_output, label)
        #errD_fake.backward()
        errD_fake.backward(retain_graph=True)
        D_G_z1 = fake_output.mean().item()
        errD = errD_real + errD_fake
        optimizerD.step()

        ############################
        # (2) Update G network: maximize log(D(G(z)))
        ###########################
        netG.zero_grad()
        netD.zero_grad()
        label.fill_(real_label)  # fake labels are real for generator cost
        # output = netD(fake)
        errG = criterion(fake_output, label)
        errG.backward()
        D_G_z2 = fake_output.mean().item()
        optimizerG.step()
```

**Comment by Zhaoyi-Yan** [comment]
It should work like this:
1. DO NOT need retain_graph
2. zero grad the netG at the begining
3. Remove zero_grad of netG and netD right before `Update G network`
```
       '''
       --------------------------------------
       Add netG.zero_grad() at the beginning
       ---------------------------------------
       '''
        netG.zero_grad()
        ############################
        # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
        ###########################
        # train with real
        netD.zero_grad()
        real_cpu = data[0].to(device)
        batch_size = real_cpu.size(0)
        label = torch.full((batch_size,), real_label, device=device)

        real_output = netD(real_cpu)
        errD_real = criterion(real_output, label)
        errD_real.backward()
        D_x = real_output.mean().item()

        # train with fake
        noise = torch.randn(batch_size, nz, 1, 1, device=device)
        fake = netG(noise)
        label.fill_(fake_label)
        # output = netD(fake.detach())
        fake_output = netD(fake)
        errD_fake = criterion(fake_output, label)
        #errD_fake.backward()
        errD_fake.backward(retain_graph=True)
        D_G_z1 = fake_output.mean().item()
        errD = errD_real + errD_fake
        optimizerD.step()

        ############################
        # (2) Update G network: maximize log(D(G(z)))
        ###########################
     # The following two lines are unnecessary!!
     #   netG.zero_grad()
     #  netD.zero_grad()
        label.fill_(real_label)  # fake labels are real for generator cost
        # output = netD(fake)
        errG = criterion(fake_output, label)
        errG.backward()
        D_G_z2 = fake_output.mean().item()
        optimizerG.step()
```

**Comment by DreamChaserMXF** [comment]
> It should work like this:
> 
> 1. DO NOT need retain_graph
> 2. zero grad the netG at the begining
> 3. Remove zero_grad of netG and netD right before `Update G network`
> 
> ```
>        '''
>        --------------------------------------
>        Add netG.zero_grad() at the beginning
>        ---------------------------------------
>        '''
>         netG.zero_grad()
>         ############################
>         # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
>         ###########################
>         # train with real
>         netD.zero_grad()
>         real_cpu = data[0].to(device)
>         batch_size = real_cpu.size(0)
>         label = torch.full((batch_size,), real_label, device=device)
> 
>         real_output = netD(real_cpu)
>         errD_real = criterion(real_output, label)
>         errD_real.backward()
>         D_x = real_output.mean().item()
> 
>         # train with fake
>         noise = torch.randn(batch_size, nz, 1, 1, device=device)
>         fake = netG(noise)
>         label.fill_(fake_label)
>         # output = netD(fake.detach())
>         fake_output = netD(fake)
>         errD_fake = criterion(fake_output, label)
>         errD_fake.backward()
>        # errD_fake.backward(retain_graph=True) # No need retain_graph
>         D_G_z1 = fake_output.mean().item()
>         errD = errD_real + errD_fake
>         optimizerD.step()
> 
>         ############################
>         # (2) Update G network: maximize log(D(G(z)))
>         ###########################
>      # The following two lines are unnecessary!!
>      #   netG.zero_grad()
>      #  netD.zero_grad()
>         label.fill_(real_label)  # fake labels are real for generator cost
>         # output = netD(fake)
>         errG = criterion(fake_output, label)
>         errG.backward()
>         D_G_z2 = fake_output.mean().item()
>         optimizerG.step()
> ```

I'm not sure but it seems that the computational graph of netG will be released when you call errD_fake.backward(). As a consequence, an exception will be raised when you call errG.backward(). More clearly, the gradients of errG need to be backpropogaed through netG but the computional graph will be lost at that time.

**Comment by Zhaoyi-Yan** [comment]
I have updated the comment.

**Comment by DreamChaserMXF** [comment]
> errD_fake.backward(retain_graph=True)

Another question is that, you said it's unnecessary to clear gradients for netD and netG between updating D and updating G. However, how do you ensure the gradients calculated in errD_fake.backward() will not be accumulated with the gradients calculated in errG.backward() ?

**Comment by shiyuanyin** [comment]
> 
> 
> I have updated the comment.
hi,
I want to ask you the target model(G model), the gradient update,  the model input is data label, out is fc, so I dont't know the Discriminator's out how to help G model gradient, I think this is no connect model???

@Zhaoyi-Yan 
@smartkiwi 
@pietern 
@ezyang

**Comment by pietern** [comment]
Please don't tag people. Questions are for the forums at https://discuss.pytorch.org/.

**Comment by shiyuanyin** [comment]
> 
> 
> Please don't tag people. Questions are for the forums at https://discuss.pytorch.org/.

OK，I have deleted

**Comment by msaroufim** [comment]
Closing since Q&A is better for forum
